---
title: "Phylogenetic Trees: Africa + Asia"
author: "Monica Mweetwa"
date: "10-11-2024"
---

```{r setup, include=FALSE}
#knitr::opts_knit$set(root.dir = normalizePath("C:/Users/Monica/Dropbox/Africa-MetaAnalysis"))
knitr::opts_knit$set(root.dir = normalizePath("~/Dropbox/Africa-MetaAnalysis")) 
```

Load libraries
```{r include=FALSE}
pacman::p_load(
  taxonomizr,      # To convert taxid IDs to taxonomy
  cluster,         # For Hierarchical Clustering
  rio,             # import/export
  here,            # relative file paths
  tidyverse,       # general data management and visualization
  tibble,          # general data management
  ape,             # to import and export phylogenetic files
  ggtree,          # to visualize phylogenetic files
  treeio,          # to visualize phylogenetic files
  ggnewscale)
```

Load data from the  search
```{r}
library(readxl)
TXID <- read_excel("All_Taxids2.xlsx")
View(TXID)
```

Retrieve database with taxonomic assignments
```{r}
prepareDatabase(getAccessions=FALSE, resume=FALSE) #this is stored in 'nameNode.sqlite'

```

Note that this might give the error ;
Error in (function (outDir = ".", url = sprintf("%s://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz",  : 
Downloaded file does not match ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz File corrupted or download ended early?

If this happens, just delete the downloaded files and try again. 

Get lineage for our dataset
```{r}
taxid <- c(TXID$taxid)
taxa_lineage <- getTaxonomy(taxid, sqlFile = "nameNode.sqlite")
```

Export taxa_lineage for 2nd tree
```{r}
taxa_lineage2 <- as.data.frame(taxa_lineage) %>%
  rownames_to_column()%>%
  mutate(Nodename = apply(., 1, function(x) tail(na.omit(x), 1))) ##To get the names from each taxonomic rank

write.csv(taxa_lineage2, "lineage2025_2.csv")
```

Rename the inner nodes that have NA in them to na1, na2, na3 e.t.c then import back to continue.

Create phylogenetic tree with edited taxonomy table
```{r}
library(readxl)
review_tree_pre <- as_tibble(read_csv("lineage2025_2_edited.csv"))
review_tree_pre <- review_tree_pre %>%
  column_to_rownames(var = "rowname")
review_tree <- makeNewick(review_tree_pre, quote = "'", excludeTerminalNAs=TRUE)
myTree1 <- ape::read.tree(text = review_tree)
```

Save tree
```{r}
ape::write.tree(myTree1, file='Tree_phytree_2025_2.nwk')
```

Edit labels
```{r}
myTree10 <- myTree1
myTree10$tip.label <- str_sub(myTree10$tip.label, end=-2)
myTree10$tip.label <- str_sub(myTree10$tip.label, start =2)

myTree10$node.label <- str_sub(myTree10$node.label, end=-2)
myTree10$node.label <- str_sub(myTree10$node.label, start =2)

```

Export tree labels and node to create metadata
```{r}
add <- as_tibble(myTree10)
write.csv(add, "Node_labels_phytree2025_2.csv")
```

Import new Metadata 
```{r}
#Import table with new values
node_lab <- read_excel("Node_labels_phytree2025_2_edited.xlsx")
node_lab$`Sub-Saharan Africa` <- ifelse(node_lab$`Sub-Saharan Africa` == 0, NA, node_lab$`Sub-Saharan Africa` )
node_lab$Other <- ifelse(node_lab$Other == 0, NA, node_lab$Other)

## Change the level of Rank
levels(as.factor(node_lab$Rank))
node_lab$Rank <- factor(node_lab$Rank, levels = c("superkingdom", "phylum", "class", "order", "family", "genus", "species", "NA"))
levels(node_lab$Rank)
#Create variable to determine branch length by using rank
node_lab$Length <- ifelse(node_lab$Rank == "species", 0.3, NA)
node_lab$Length <- ifelse(node_lab$Rank == "genus", 0.3, node_lab$Length)
node_lab$Length <- ifelse(node_lab$Rank == "family", 0.3, node_lab$Length)
node_lab$Length <- ifelse(node_lab$Rank == "order", 0.3, node_lab$Length)
node_lab$Length <- ifelse(node_lab$Rank == "class", 0.3, node_lab$Length)
node_lab$Length <- ifelse(node_lab$Rank == "phylum", 0.3, node_lab$Length)
node_lab$Length <- ifelse(node_lab$Rank == "superkingdom", 0.3, node_lab$Length)
node_lab$Length <- ifelse(is.na(node_lab$Rank), 0, node_lab$Length)

## Change the level of Rank
levels(as.factor(node_lab$Rank))
node_lab$Rank <- factor(node_lab$Rank, levels = c("superkingdom", "phylum", "class", "order", "family", "genus", "species", "NA"))
levels(node_lab$Rank)

#Join metadata with phylotree
y33<- left_join(add, node_lab, by = 'label')

y33$label <- ifelse(y33$node == 209 , "Center", y33$label)
y34 <- as.treedata(y33)

y34
```

plot - Combine Rank and Counts
Color of edges is discrete
1. All
```{r}
#install.packages("viridis")
library(viridis)
okabe <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

p1 <- ggtree(y34, layout='circular', branch.length = 'Length') +
  geom_tree(aes(color = `Grand Total`), size=3, inherit.aes = TRUE) +
  #scale_colour_gradient(low = "purple", mid = "yellow", high = "green","Number of \nStudies", na.value = NA) +
  scale_colour_viridis_c(name = "Number of \nStudies", option = "B", na.value = NA, begin = 0.3, limits = c(1,20)) +
  geom_tiplab2(aes(label= ShortName), color = "black", align = F, size=8, hjust= -0.1) +
  geom_tiplab2(aes(label= ShortName2), color = "black", align = T, linetype= "dotted", size=8, hjust= 0.1, na.translate=FALSE) +
  new_scale_color() + 
  geom_nodepoint(aes(color = Rank), size=4, show.legend = TRUE) +
  geom_tippoint(aes(color = Rank), size=4) + 
  scale_colour_manual(values = c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "red"), na.value = NA, name = "Rank", na.translate=FALSE) +
  theme(legend.position = "left",
        legend.text = element_text(size = 8),
        legend.title = element_text(size=8, face="bold"),
        legend.key = element_blank(),
        legend.background = element_blank(),
        legend.box.background = element_rect(colour = "black", linewidth = 1),
        plot.title = element_text(size = 24, hjust = -0.2),
        panel.background = element_blank(),
        plot.background = element_blank(), 
        panel.grid = element_blank()) + 

  geom_rootpoint(colour = "red", size=3, shape=8)

library(Cairo)
Cairo(file = "Figures/Combined/phytree_all_2025_May.png", 
      type = "png", 
      units = "px",
      width = 12000, 
      height = 10000, dpi = 350, bg = "transparent")
p1 + theme(legend.position = "none")
invisible(dev.off())
```

2. Africa
```{r}
library(viridis)
okabe <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
p2 <- ggtree(y34, layout='circular', branch.length = 'Length') +
  geom_tree(aes(color = `Sub-Saharan Africa`),  size=3, inherit.aes = TRUE) +
  scale_colour_viridis_c(name = "Number of \nStudies", option = "B", na.value = NA, begin = 0.3, limits = c(1,20)) +
  new_scale_color() + 
  geom_nodepoint(aes(color = Rank), size=4, show.legend = TRUE) +
  geom_tippoint(aes(color = Rank), size=4) + 
  scale_colour_manual(values = c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "red"), na.value = NA, name = "Rank", na.translate=FALSE) +
  theme(legend.position = "left",
        legend.text = element_text(size = 8),
        legend.title = element_text(size=8, face="bold"),
        legend.key = element_blank(),
        legend.background = element_blank(),
        legend.box.background = element_rect(colour = "black", linewidth = 1),
        plot.title = element_text(size = 24, hjust = -0.2),
        panel.background = element_blank(),
        plot.background = element_blank(), 
        panel.grid = element_blank()) + 
  geom_rootpoint(colour = "red", size=3, shape=8)
```

3. Other LMICs
```{r}
library(viridis)
okabe <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
p3 <- ggtree(y34, layout='circular', branch.length = 'Length') +
  geom_tree(aes(color = Other),  size=3, inherit.aes = TRUE) +
  scale_colour_viridis_c(name = "Number of \nStudies", option = "B", na.value = NA, begin = 0.3, limits = c(1,20)) +
  new_scale_color() + 
  geom_nodepoint(aes(color = Rank), size=4, show.legend = TRUE) +
  geom_tippoint(aes(color = Rank), size=4) + 
  scale_colour_manual(values = c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "red"), na.value = NA, name = "Rank", na.translate=FALSE) +
  theme(legend.position = "left",
        legend.text = element_text(size = 8),
        legend.title = element_text(size=8, face="bold"),
        legend.key = element_blank(),
        legend.background = element_blank(),
        legend.box.background = element_rect(colour = "black", linewidth = 1),
        plot.title = element_text(size = 24, hjust = -0.2),
        panel.background = element_blank(),
        plot.background = element_blank(), 
        panel.grid = element_blank()) + 
  geom_rootpoint(colour = "red", size=3, shape=8)
```

Save figures for combination in Gimp
```{r}
ggsave("Figures/Combined/CombinePhyplot_All.png", plot = p1+ theme(legend.position = "none"), dpi = 350, device = "png",  
      units = "px",
      width = 12000, 
      height = 10000)
ggsave("Figures/Combined/CombinePhyplot_Africa.png", plot = p2+ theme(legend.position = "none"), dpi = 600, device = "png", width = 24, height = 24, units = "in")
ggsave("Figures/Combined/CombinePhyplot_Other.png", plot = p3+ theme(legend.position = "none"), dpi = 600, device = "png", width = 24, height = 24, units = "in")

combplot <- ggarrange(p1, ggarrange(p2+ theme(legend.position = "none"), 
                        p3+ theme(legend.position = "none"), 
                        ncol = 1, labels = c("(b)", "(c)")),
          ncol = 2, nrow = 1, labels = c("(a)"), common.legend = T, legend = "bottom")

ggsave("Figures/Combined/CombinePhyplot_Region.png", plot = last_plot(), dpi = 600, device = "png", width = 30, height = 12, units = "in")

```

save legend as well
```{r}
grobs <- ggplotGrob(p1)$grobs
legend <- grobs[[which(sapply(grobs, function(x) x$name) == "guide-box")]]

#Save legend
Cairo(file = "Figures/Combined/MainPhytree-fdr-legend_May.png", 
      type = "png", 
      units = "px",
      width = 9000, 
      height = 10000, dpi = 350, bg = "transparent")
b1 <- plot(legend)
b1 + theme(legend.background = element_blank(),
        legend.box.background = element_rect(colour = "black", linewidth = 1),
        plot.title = element_text(size = 24, hjust = -0.2),
        panel.background = element_blank(),
        plot.background = element_blank(), 
        panel.grid = element_blank())
ggsave("Figures/Combined/MainPhytree.png", plot = p3+ theme(legend.position = "none"), dpi = 600, device = "png",  
      units = "px",
      width = 9000, 
      height = 10000, dpi = 350)

```


Sequencing Type
```{r}
library(viridis)
okabe <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
#A. Targeted sequencing
p4 <- ggtree(y34, layout='circular', branch.length = 'Length') +
  geom_tree(aes(color = `Targeted Sequencing`),  size=3, inherit.aes = TRUE) +
  scale_colour_viridis_c(name = "Number of \nStudies", option = "B", na.value = NA, begin = 0.3, limits = c(1,34)) +
  geom_tiplab2(aes(label= ifelse(is.na(`Targeted Sequencing`), NA,ShortName)), color = "black", align = F, size=8, hjust= -0.1) +
  geom_tiplab2(aes(label= ifelse(is.na(`Targeted Sequencing`), NA,ShortName2)), color = "black", align = T, linetype= "dotted", size=8, hjust= 0.1,
  na.translate=FALSE) +
  new_scale_color() + 
  geom_nodepoint(aes(color = Rank), size=4, show.legend = TRUE) +
  geom_tippoint(aes(color = Rank), size=4) + 
  scale_colour_manual(values = c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "red"), na.value = NA, name = "Rank", na.translate=FALSE) +
  theme(#legend.position = c(1.2, 0.5),
        legend.position = "left",
        legend.text = element_text(size = 8),
        legend.title = element_text(size=8, face="bold"),
        legend.key = element_blank(),
        legend.background = element_blank(),
        legend.box.background = element_rect(colour = "black", linewidth = 1),
        plot.title = element_text(size = 24, hjust = -0.2),
        panel.background = element_blank(),
        plot.background = element_blank(), 
        panel.grid = element_blank()) + 
  geom_rootpoint(colour = "red", size=3, shape=8)

#B. Whole genome sequencing
p5 <- ggtree(y34, layout='circular', branch.length = 'Length') +
  geom_tree(aes(color = `Whole genome sequencing`),  size=3, inherit.aes = TRUE) +
  scale_colour_viridis_c(name = "Number of \nStudies", option = "B", na.value = NA, begin = 0.3, limits = c(1,34)) +
  geom_tiplab2(aes(label= ifelse(is.na(`Whole genome sequencing`), NA,ShortName)), color = "black", align = F, size=8, hjust= -0.1) +
  geom_tiplab2(aes(label= ifelse(is.na(`Whole genome sequencing`), NA,ShortName2)), color = "black", align = T, linetype= "dotted", size=8, hjust= 0.1, na.translate=FALSE) +
  new_scale_color() + 
  geom_nodepoint(aes(color = Rank), size=4, show.legend = TRUE) +
  geom_tippoint(aes(color = Rank), size=4) + 
  scale_colour_manual(values = c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "red"), na.value = NA, name = "Rank", na.translate=FALSE) +
  theme(#legend.position = c(1.2, 0.5),
        legend.position = "left",
        legend.text = element_text(size = 8),
        legend.title = element_text(size=8, face="bold"),
        legend.key = element_blank(),
        legend.background = element_blank(),
        legend.box.background = element_rect(colour = "black", linewidth = 1),
        plot.title = element_text(size = 24, hjust = -0.2),
        panel.background = element_blank(),
        plot.background = element_blank(), 
        panel.grid = element_blank()) + 
  geom_rootpoint(colour = "red", size=3, shape=8)

#save
ggsave("Figures/Combined/CombinePhyplot_TargetedSeg.png", plot = p4+ theme(legend.position = "none"), dpi = 350, device = "png",  
      units = "px",
      width = 12000, 
      height = 10000)
ggsave("Figures/Combined/CombinePhyplot_WholeGenSeq.png", plot = p5+ theme(legend.position = "none"), dpi = 350, device = "png",  
      units = "px",
      width = 12000, 
      height = 10000)

#Save legend
grobs <- ggplotGrob(p4)$grobs
legend <- grobs[[which(sapply(grobs, function(x) x$name) == "guide-box")]]
b1 <- plot(legend)
b1 + theme(legend.background = element_blank(),
        legend.box.background = element_rect(colour = "black", linewidth = 1),
        plot.title = element_text(size = 24, hjust = -0.2),
        panel.background = element_blank(),
        plot.background = element_blank(), 
        panel.grid = element_blank())
ggsave("Figures/Combined/SequencingMethod_legend.png", plot = c1, dpi = 600, device = "png", units = "px",
      width = 9000, 
      height = 10000,
      limitsize = FALSE)

```

Fishers exact test
```{r}
library(tidyverse)

node_lab <- read_excel("Tables/Phytree_tabs/Node_labels_phytree2025_2_edited.xlsx")

node_lab2 <- node_lab %>%
  filter(`Grand Total` != 0) %>%
  filter(`Grand Total` >= 5)%>%
  mutate(`Sub-Saharan Africa` = as.numeric(`Sub-Saharan Africa`),
         Other = as.numeric(Other),
         `Targeted Sequencing` = as.numeric(`Targeted Sequencing`),
         `Whole genome sequencing` = as.numeric(`Whole genome sequencing`)) %>%
  replace_na(list(`Sub-Saharan_Africa` = 0, Other = 0, `Targeted Sequencing` = 0, `Whole genome sequencing` = 0))
  
#pivot wider
node_lab3_region <- node_lab2 %>%
  select(label, `Sub-Saharan Africa`, Other) %>%
  group_by(label) %>%
  pivot_longer(cols = c(`Sub-Saharan Africa`, Other), names_to = "Region", values_to = "num")
node_lab3_method <- node_lab2 %>%
  select(label, `Targeted Sequencing`, `Whole genome sequencing`) %>%
  group_by(label) %>%
  pivot_longer(cols = c(`Targeted Sequencing`, `Whole genome sequencing`), names_to = "Method", values_to = "num")

#Draw histogram
library(ggplot2)
library(forcats)
#Region
ggplot(node_lab3_region, aes(fill = Region, x = fct_reorder(label, num), y = num)) +
  geom_bar(stat = "identity", position = position_dodge()) + 
  geom_text(aes(label = num), hjust= 1.6, color = "black", position = position_dodge(0.9)) +
  labs(x = "Top Microbes Detected",
       y = "Number of studies") +
  theme(legend.position = "none") +
  coord_flip() +
  theme_classic()
ggsave("Figures/Combined/DistrbutionOfTopMicrobes.png", plot = last_plot(), dpi = 600, device = "png", width = 6, height = 6, units = "in")

#fishers test
node_lab4 <- node_lab%>%
  filter(!is.na(`Grand Total`)) %>%
  filter(`Grand Total` > 0)%>%
  mutate(`Sub-Saharan Africa` = as.numeric(`Sub-Saharan Africa`),
         Other = as.numeric(Other)) %>%
  select(label, `Sub-Saharan Africa`, Other) %>%
  replace_na(list(`Sub-Saharan Africa` = 0, Other = 0))%>%
  group_by(label) %>%
  pivot_longer(cols = c(`Sub-Saharan Africa`, Other), names_to = "Region", values_to = "num")%>%
  replace_na(list(num = 0))%>%
  mutate(num2 = ifelse(Region == "Other", 18 - num, 16 - num))

#Create matrix tables for each microbe
contingency_matrices <- node_lab4 %>%
  group_by(label) %>%
  summarise(
    matrix = list(matrix(
      c(
        sum(num[Region == "Other"]), sum(num[Region == "Sub-Saharan Africa"]),
        sum(num2[Region == "Other"]), sum(num2[Region == "Sub-Saharan Africa"])
      ),
      nrow = 2,
      byrow = TRUE
    ))
  ) %>%
  pull(matrix)

set.seed(123) # For reproducibility

# Function to perform permutations for a single table
permute_p_value <- function(tbl, n_perms = 1000) {
  observed_p <- fisher.test(tbl)$p.value
  
  # Generate null distribution
  null_p <- replicate(n_perms, {
    # Permute table (shuffle the counts)
    perm_tbl <- tbl
    perm_tbl[, 1] <- sample(tbl[, 1])
    perm_tbl[, 2] <- sample(tbl[, 2])
    fisher.test(perm_tbl)$p.value
  })
  
  # Calculate empirical p-value
  mean(null_p <= observed_p)
}

# Apply permutation test to each table
empirical_p_values <- sapply(contingency_matrices, permute_p_value)

# Results
fish_result <- node_lab4 %>%
  group_by(label) %>%
  summarise(
    p_value = {
      contigency_table <- matrix(c(
        num[Region == "Other"], num[Region == "Sub-Saharan Africa"],
        num2[Region == "Other"], num2[Region == "Sub-Saharan Africa"]
      ), nrow = 2)
      fisher.test(contigency_table)$p.value
    }
  )
```


done!